### PR TITLE
[BugFix][CONV1D] Align tensor layout with manifest

### PR DESCRIPTION
**Pull request by RMLYC**
## Summary

This PR aligns Conv1d runtime tensor layout with the manifest-declared NCL semantics and adds a dedicated pointwise Conv1d fast path.

- Updates Conv1d kernel inputs/outputs to use `x: (N, C_in, L)` and `out: (N, C_out, L_out)`.
- Reworks the generic Conv1d tile grid to use `(L tile, C_out tile, N)` so the L dimension remains contiguous instead of forcing NCL data into the previous NLC-optimized flattened structure.
- Flattens generic Conv1d weight as `(C_out, K * C_in)` before the TileLang kernel so the kernel can use a contiguous shared-memory copy for weights.
- Adds an input tile fast path for fully in-bounds generic Conv1d tiles to reduce boundary checks on common interior tiles.
- Adds `Conv1dPointwiseKernel` as a peer kernel to `Conv1dKernel`, similar to `Conv2d1x1Kernel`.
- Moves pointwise dispatch to `Conv1dFwdOp`, matching the `Conv2dOp` pattern for selecting `Conv2d1x1Kernel` vs `Conv2dKernel`.

## Test plan

- `python -m py_compile tileops/ops/convolution.py tileops/kernels/convolution.py tileops/kernels/__init__.py`
- `python -m ruff check tileops/ops/convolution.py tileops/kernels/convolution.py tileops/kernels/__init__.py`
- `git diff --check`
- `python scripts/validate_manifest.py --check-op Conv1dFwdOp --levels schema,signature`
- `python scripts/validate_manifest.py --check-op Conv1dBiasFwdOp --levels schema,signature`
- `TMPDIR=/home/lyc/Project/TileOps-workspace2/.tmp TILELANG_CLEANUP_TEMP_FILES=1 python -m pytest tests/ops/test_convolution.py -k 'pointwise' -vvs`
- `TMPDIR=/home/lyc/Project/TileOps-workspace2/.tmp TILELANG_CLEANUP_TEMP_FILES=1 python -m pytest tests/ops/test_convolution.py -k 'conv1d and not pointwise' -vvs`
- `CUDA_VISIBLE_DEVICES=1 TMPDIR=/home/lyc/Project/TileOps-workspace2/.tmp TILELANG_CLEANUP_TEMP_FILES=1 python -m pytest benchmarks/ops/bench_convolution.py -k 'convtasnet-pointwise-k1-s1-fp16' -vvs`

## Benchmark

Full Conv1d benchmark on NVIDIA H200, comparing against the parent before the layout change:

| case | before TileOps ms | current TileOps ms | change |
| --- | ---: | ---: | ---: |
| convtasnet-pointwise-k1-s1-fp16 | 0.1898 | 0.2448 | +29.0% |
| seanet-k3-s1-fp16 | 0.0175 | 0.0213 | +21.7% |
| audio-downsample-k5-s2-fp16 | 0.0160 | 0.0199 | +24.4% |
| seanet-stem-k7-s1-fp16 | 0.0431 | 0.0618 | +43.4% |
| sequence-downsample-k3-s2-bf16 | 0.0097 | 0.0105 | +8.2% |
| seanet-k3-s1-d2-fp16 | 0.0174 | 0.0192 | +10.3% |

Pointwise rerun after adding `Conv1dPointwiseKernel`, on GPU1 with SM clock checked at 1830 MHz:

| case | TileOps ms | Torch ms | TileOps TFLOPS | bandwidth TB/s |
| --- | ---: | ---: | ---: | ---: |
| convtasnet-pointwise-k1-s1-fp16 | 0.2448 | 0.5180 | 137.0594 | 0.8042 |

Raw local reports:

- `output_mid/conv1d_manifest/profile_before_24a036c.log`
- `output_mid/conv1d_manifest/profile_current_full_gpu0.log`
- `output_mid/conv1d_manifest/full_benchmark_comparison.md`
- `output_mid/conv1d_manifest/profile_conv1d_pointwise_op_dispatch_gpu1.log`

## Regression

This PR intentionally prioritizes manifest layout correctness over preserving the previous NLC-oriented Conv1d optimization. Current generic Conv1d performance is slower than the previous implementation because:

- The corrected NCL layout no longer matches the original flattened NLC-optimized traversal.
- The generic Conv1d path now carries additional index mapping and boundary handling for stride, padding, and dilation.
- `weight.permute(0, 2, 1).contiguous()` is currently done in the Python forward path for generic Conv1d, which adds end-to-end overhead.
- Reverting to a 2D flattened `(N * L)` grid reintroduces per-element `m_idx // out_l` and `m_idx % out_l` mapping and severely regresses K=7/S=1, so this PR keeps the 3D grid.

The largest previous regression was K=1/S=1 pointwise Conv1d. This PR now routes that case through `Conv1dPointwiseKernel`, which removes generic stride/padding/dilation mapping from the TileLang kernel and lowers the pointwise result from the earlier corrected-layout measurement of `0.3015 ms` to `0.2448 ms`.

Follow-up directions:

- Further tune pointwise Conv1d to close the remaining gap against the old NLC-oriented baseline.
- Add or cache a weight-packed representation to reduce repeated host-side weight layout conversion.
- Continue tuning K=3/S=1 and K=7/S=1 after the correctness-aligned generic kernel is merged.
- Use NCU on pointwise and K=3/K=7 cases to separate index overhead, memory load efficiency, and weight packing cost.

### CODE REVIEW

File: tileops/kernels/convolution.py [hunk 1]
```diff
@@ -67,72 +67,82 @@ def _conv1d_func(
...
+                    tile_ol_start = bx * block_n
+                    tile_ol_end = tile_ol_start + block_n - 1
+                    tile_input_start = tile_ol_start * stride_l - pad_l
+                    tile_input_end = tile_ol_end * stride_l + (kernel_l - 1) * dilation_l - pad_l
+                    tile_full = (
+                        (tile_ol_end < out_l)
+                        & (tile_input_start >= 0)
+                        & (tile_input_end < l_in)
+                        & ((k_iter + 1) * block_k <= k_total)
+                    )
```
[the remark this answers is not in the document]
Comment (RMLYC) [reply tileops/kernels/convolution.py]: Updated as suggested: `tile_ol_start`, `tile_ol_end`, `tile_input_start`, `tile_input_end`, and the spatial boundary check have been hoisted out of the `T.Pipelined` loop. The loop now only keeps the `block_k` boundary predicate that depends on `k_iter`.

Commit: 2989040 `[Perf][CONV1D] Hoist tile boundary checks`

Validation:
- `python -m py_compile tileops/kernels/convolution.py`: passed
- `python -m pytest tests/ops/test_convolution.py -k 'conv1d' -vvs`: 13 passed
- `CUDA_VISIBLE_DEVICES=1 python -m pytest benchmarks/ops/bench_convolution.py -k 'audio-downsample-k5-s2-fp16 or seanet-stem-k7-s1-fp16' -vvs`: 2 passed

Performance is effectively unchanged:
- audio-downsample-k5-s2-fp16: `0.0199 -> 0.0202 ms`
- seanet-stem-k7-s1-fp16: `0.0618 -> 0.0632 ms`

**Review by superAngGao — CHANGES_REQUESTED** [review]
Thanks for fixing the Conv1d layout mismatch and for documenting the performance tradeoff. I think the NCL semantic change itself is the right direction, but I found one runtime regression that should block this before merge.

P1: Conv1d float32 now fails at runtime on H200 with the new default config. Both the generic and pointwise paths use the SM90 default {block_m=64, block_n=128, block_k=128, num_stages=3}. For float32 this asks for about 294912 bytes of dynamic shared memory, and launching the kernel fails with:

    InternalError Failed to set the allowed dynamic shared memory size to 294912

I reproduced this on the PR for both kernel_size=1 and kernel_size=3. The same generic Conv1d float32 case runs on the parent commit, whose default config used block_k=64. Since the manifest still declares Conv1d input dtype as float32 | float16 | bfloat16, this is a user-visible regression rather than just a perf issue. Please make default_config dtype/shared-memory aware, or at least use a smaller float32 default, and add a float32 smoke case.

Two smaller follow-ups I would also like to see:

- Please add an explicit dispatch test for the pointwise fast path. test_conv1d_dispatches_kernel still only verifies the generic Conv1dKernel path; there is no assertion that k=1/stride=1/padding=0/dilation=1 selects Conv1dPointwiseKernel.
- Please update the Conv1d manifest source.kernel_map entries to include conv1d_pointwise_kernel: Conv1dPointwiseKernel. The implementation now has two dispatch targets, but the manifest still documents only conv1d_kernel.

I also ran some local checks: conv1d smoke tests pass, and fp16/bf16 custom non-aligned generic/pointwise cases match torch. So aside from the float32 default-config regression, the correctness direction looks good to me.

**Comment by michaelwithu** [comment]
## Review

Interface alignment with `torch.nn.functional.conv1d` (NCL) is correct per the manifest trust-model. Performance regression is the inherent cost of moving channel off the innermost dimension — acceptable given the contract fix. LGTM after one change.

### Request: cache weight permute

`Conv1dKernel.forward` runs `weight.permute(0, 2, 1).contiguous().view(...)` on every call — allocates + memcpy each time. Weight is invariant at inference; move this to `__init__`. Free win for all shapes, especially sub-0.1ms cases.

### Question: pointwise GEMM convention

`Conv2d1x1Kernel` uses `T.gemm(data, weight, out, transpose_B=True)` so weight's c_in stays contiguous. The new `Conv1dPointwiseKernel` uses `T.gemm(weight, data, out)` without transpose_B. Intentional divergence, or worth aligning?

### Follow-up suggestions (not blocking)

- Hoist `kw = k_idx // c_in`, `ci = k_idx % c_in` out of `T.Parallel(block_k, block_n)` — they only depend on `i`, saving `block_n×` div/mod ops.
- Peel last k_iter to eliminate the per-iter `if tile_full` branch from the common path.
- NCU on K=3/K=7 to isolate index arithmetic vs non-coalesced gather vs occupancy.

**Comment by RMLYC** [comment]
Thanks for the detailed review.

I investigated the fp32 Conv1d path and confirmed that the current TileLang `T.gemm` lowering does not provide full-fp32 multiply for this kernel. For the `smoke-tcn-k3-s1-fp32` case, the generated CUDA contains:

```cpp
tl::mma_sync<
  tl::DataType::kTensorFloat32,
  tl::DataType::kTensorFloat32,
  tl::DataType::kFloat32,
  16, 8, 8, false, true
>(...)
```

So fp32 inputs are lowered through TF32 MMA with fp32 accumulation. This explains the observed precision loss and means the current Conv1d kernel cannot satisfy strict fp32 correctness requirements. The TileLang-side limitation is tracked in https://github.com/tile-ai/tilelang/issues/859.

Given that, I removed fp32 from the Conv1d manifest signature instead of adding a widened tolerance or a fp32 smoke test. Conv1d is now declared for `float16 | bfloat16` only until TileLang supports a full-fp32 `T.gemm` path.

I kept the two actionable non-fp32 follow-ups:

- Added smoke dispatch coverage for both generic `Conv1dKernel` and pointwise `Conv1dPointwiseKernel`.
- Updated the Conv1d manifest `source.kernel_map` entries to include `conv1d_pointwise_kernel: Conv1dPointwiseKernel` for both `Conv1dFwdOp` and `Conv1dBiasFwdOp`.

Commit: 2618e07 `[BugFix][CONV1D] Update pointwise dispatch coverage`

Validation:

- `python -m py_compile tests/ops/test_convolution.py`: passed
- `python -m ruff check tests/ops/test_convolution.py`: passed
- `git diff --check`: passed
- `TMPDIR=/home/lyc/Project/TileOps-workspace2/.tmp TILELANG_CLEANUP_TEMP_FILES=1 python -m pytest tests/ops/test_convolution.py -k 'test_conv1d_dispatches_kernel or pointwise' -vvs`: 3 passed
- `python scripts/validate_manifest.py --check-op Conv1dFwdOp --levels schema,signature`: passed
- `python scripts/validate_manifest.py --check-op Conv1dBiasFwdOp --levels schema,signature`: passed

**Review by superAngGao — DISMISSED** [review]
LGTM.

**Comment by RMLYC** [comment]
Thanks for the review. I addressed the blocking weight-permute request in commit a085e4f `[Perf][CONV1D] Cache flattened generic weights`.

`Conv1dKernel` now keeps a single-entry cache for the flattened generic weight tensor. It reuses the cached `weight.permute(0, 2, 1).contiguous().view(...)` result when the same weight tensor is passed again and its tensor `_version` has not changed. If a different weight tensor is used, or the original tensor is modified in-place, the cache is refreshed.

On the pointwise GEMM convention question: the current `Conv1dPointwiseKernel` uses `(C_out, C_in) @ (C_in, L)` intentionally to keep the output tile/layout aligned with the Conv1d NCL layout work in this PR. I am leaving that convention unchanged in this PR.

The suggested `kw/ci` hoist, last-k peel, and K=3/K=7 NCU analysis are useful follow-ups, but I am not including them here to keep this PR focused on the manifest layout fix and the pointwise dispatch path.

Validation:

- `python -m py_compile tileops/kernels/convolution.py`: passed
- `python -m ruff check tileops/kernels/convolution.py`: passed
- `git diff --check`: passed
- `TMPDIR=/home/lyc/Project/TileOps-workspace2/.tmp TILELANG_CLEANUP_TEMP_FILES=1 python -m pytest tests/ops/test_convolution.py -k 'conv1d' -vvs`: 14 passed

Pointwise benchmark was also rerun. GPU1 was occupied by a background Python process and produced unstable numbers, so I also reran on idle GPU0 for a sanity check:

- GPU0 `convtasnet-pointwise-k1-s1-fp16`: TileOps `0.2486 ms`, Torch `0.5011 ms`
- best config: `block_m=128, block_n=256, block_k=64, num_stages=2, threads=256`

The weight cache only affects the generic Conv1d path; pointwise dispatch and kernel code are unchanged by this commit.

**Review by lcy-seso — APPROVED** [review]
